### ISSUE TITLE
as_string() function for psycopg2.sql module requires a "context" (connection or cursor).

### ISSUE THREAD

**Issue by zlex7** [psycopg/psycopg2#747]
The as_string(conn) function from the psycopg2.sql module is used to convert a sql.SQL object to its underlying string form. This is causing me a few difficulties when trying to unit test some functions that use the sql.SQL module, as I do not want to spin up a database just to unit test. Yes, one solution could be to try to mock the connection or just test against the SQL objects before converting them to a string. However, is there any particular reason that the as_string() function necessitates a context? Could it not be treated as completely independent from any particular database?

**Comment by fogzot** [comment]
Unfortunately, no. The string should have the correct encoding and to know what encoding the backend expects, we need a connection.

We _could_ use a default encoding, probably UNICODE, if the connection is null or add an `encoding` argument to `as_string()`. It is a bit of work and should be done in a backward compatible way. Patches are welcome.

**Comment by dvarrazzo** [comment]
Not only certain results depend on the encoding. Several functions are just wrapper to the underlying libpq library, which takes a libpq connection object as argument: psycopg offers to wrap them and there is no plan for us to re-implement them in pure Python. So no, this won't happen.

**Comment by zlex7** [comment]
Ok, thank you for explaining. Do you have any suggestions about how to workaround this or mock a database connection so I don't need to run a PostgreSQL instance simply to write some queries?

**Comment by zlex7** [comment]
I am looking at the test_sql.py tests for the sql module, and it looks like it uses an actual connection to compare the string forms. However, this doesn't really seem like something I would want to do in order to run lots of fast, low overhead unit tests.

**Comment by fogzot** [comment]
Mocking thr connection is a bad idea. What I'd do is to abstract the db access in a XyzStore class (or classes) and then mock it to test the rest of my code.

Testing the sql is an integration test and should be done on a real db with the exact version used in ptoduction. PostgreSQL is known for introducing forward incompatible changes, especially when automatic type casting of parameters is involved.

**Comment by zlex7** [comment]
I think you might misunderstand my situation, probably because I haven't explained it well. So, I do already have an XyzStore class (I just call it PostgreSQL) to abstract the connection. What I am attempting to test is a JSON-to-SQL parser that I have written, that uses psycopg2.sql to actually build the SQL query (for escaping/security purposes). I simply want to test if the parser is working correctly by feeding it specific JSON examples and confirming that it is parsing and converting to SQL correctly. However, it is very hard (and ugly) to compare the resulting SQL I have generated without being able to actually convert it to a string at the end (as I would have to directly compare against Composed, SQL, Identifier, etc. objects). Lastly, I do not want to start a database simply to test whether the SQL query I have dynamically built looks like how it should. I don't believe the encoding matters in my case, only that the formatting of the queries is consistent across tests. So basically all I want to do is have some way of getting the underlying string representation of my SQL query without making a database connection. I'm not sure that mocking is even possible, seeing that the code checks whether the given context is a Cursor or Connection object. Do you have any idea what I could do? Thanks for your help!

**Comment by zlex7** [comment]
I could be completely wrong here, but it appears that the only references to the given context inside of the exposed as_string() functions are 
```if hasattr(a, 'prepare'):
            a.prepare(conn)```
and
```if sys.version_info[0] >= 3 and isinstance(rv, bytes):
            rv = rv.decode(ext.encodings[conn.encoding])```
and
```def as_string(self, context):
           return ext.quote_ident(self._wrapped, context)```

In the first and second case it appears that it is not absolutely necessary to supply a connection (can not have 'prepare' attribute for first case and provide just an encoding for the second). However, after looking deeper, the third case seems to be what is making this function require a connection:

The underlying function called for quote_ident() is actually psyco_quote_ident(), which then uses the (connectionObject* conn) struct only twice.

These two calls are to

```PyObject *
conn_text_from_chars(connectionObject *self, const char *str)
{
    return psycopg_text_from_chars_safe(str, -1, self ? self->pydecoder : NULL);
}
```

and

```char *
psycopg_escape_identifier(connectionObject *conn, const char *str, Py_ssize_t len)
{
    char *rv = NULL;

    if (!conn || !conn->pgconn) {
        PyErr_SetString(InterfaceError, "connection not valid");
        goto exit;
    }

    if (len < 0) { len = strlen(str); }

    rv = PQescapeIdentifier(conn->pgconn, str, len);
    if (!rv) {
        char *msg;
        msg = PQerrorMessage(conn->pgconn);
        if (!msg || !msg[0]) {
            msg = "no message provided";
        }
        PyErr_Format(InterfaceError, "failed to escape identifier: %s", msg);
    }

exit:
    return rv;
}```

In the first case here, it appears that a non-null connection is not required (as it defaults to NULL), which is confirmed when looking at the underlying  psycopg_text_from_chars_safe() function.

in the second case, the connection instance is only really needed in one line, a call to PQEscapeIdentifier(), which is part of the libpq library, as mentioned earlier by @dvarrazzo.

Mr. Varrazzo also mentioned that this function (taken from the libpq library) relies on taking a connection object as a parameter.

So, PQEscapeIdentifier() calls a function PQEscapeInternal(), which then only references the given connection object four times, twice to access the given encoding, and twice to simply change an error message variable.

**Relevant Locations**

```charlen = pg_encoding_mblen(conn->client_encoding, s);```

```int i = pg_encoding_mblen(conn->client_encoding, s);```

```printfPQExpBuffer(&conn->errorMessage,
						  libpq_gettext("incomplete multibyte character\n"));```

```printfPQExpBuffer(&conn->errorMessage,
						  libpq_gettext("out of memory\n"));```


So, to finally get to my point, is there any possibility of simply supplying a connection object that only has values set for the encoding (which is all that matters in this case)? This could, for example, be exposed through a mock_connection() function in the higher-level python module that would supply a "connection" for exactly the type of purpose I want to use it for. Would this be feasible?

**Comment by zlex7** [comment]
Fair enough, and thanks for the response. However, if I get it working, can I submit a pull request for this issue?

**Comment by ackerleytng** [comment]
@zlex7 I felt that I needed this too.

For my rather basic test, I ended up using a helper test function that was able to form the query

```
def _join_seq(seq):
    parts = str(seq).split("'")
    return "".join([p for i, p in enumerate(parts) if i % 2 == 1])
```

and then I used this with

```
_join_seq(query.seq)
```

instead of 

```
query.as_string(conn)
```

**Comment by Qveshn** [comment]
### My solution
code:
```
from psycopg2 import sql


def as_string(composable):
    if isinstance(composable, sql.Composed):
        return ''.join([as_string(x) for x in composable])
    elif isinstance(composable, sql.SQL):
        return composable.string
    else:
        rv = sql.ext.adapt(composable._wrapped).getquoted()
        return rv.decode() if isinstance(rv, bytes) else rv


query = sql.SQL("insert into mytable VALUES ({});") \
    .format(sql.SQL(', ').join(map(sql.Literal, ["any 'text'", 1.2e-1, datetime.date.today()])))

print(as_string(query))
print(as_string(sql.Literal("1")))
print(as_string(sql.Literal(1e-2)))
print(as_string(sql.Literal(datetime.datetime.now())))
```
output:
```
insert into mytable VALUES ('any ''text''', 0.12, '2021-05-09'::date);
'1'
0.01
'2021-05-09T00:01:02.629068'::timestamp
```

**Comment by Qveshn** [comment]
```
print(as_string(sql.Literal("€".encode())))
```
😄

**Comment by Qveshn** [comment]
Look, I'm new to python and postgresql 😄 
Please tell me what to expect from "€"?
What should the result look like?

**Comment by Qveshn** [comment]
### My solution 2 😃 
code:
```
from psycopg2 import sql
import psycopg2.extensions


def as_string(composable):
    if isinstance(composable, sql.Composed):
        return ''.join([as_string(x) for x in composable])
    elif isinstance(composable, sql.SQL):
        return composable.string
    else:
        rv = sql.ext.adapt(composable._wrapped)
        if isinstance(rv, psycopg2.extensions.QuotedString): rv.encoding = "utf-8"
        rv = rv.getquoted()
        return rv.decode("utf-8") if isinstance(rv, bytes) else rv


query = sql.SQL("insert into mytable VALUES ({});") \
    .format(sql.SQL(', ').join(map(sql.Literal, ["any 'text' in 日本語 and even €", 1.2e-1, datetime.date.today()])))

print(as_string(query))
print(as_string(sql.Literal("1")))
print(as_string(sql.Literal(1e-2)))
print(as_string(sql.Literal(datetime.datetime.now())))
print(as_string(sql.Literal("ภาษาใดก็ได้")))
```
output:
```
insert into mytable VALUES ('any ''text'' in 日本語 and even €', 0.12, '2021-05-09'::date);
'1'
0.01
'2021-05-09T00:36:59.329131'::timestamp
'ภาษาใดก็ได้'
```

**Comment by Qveshn** [comment]
>>The solution depends on the connection encoding.
Hmm... I'm new to postgres.
"connection encoding" - are you talking about client connection some attribute or database charset encoding?

My task is to prepare statements and save them somewhere (for example in "file.sql").
Then I am going to execute that file via connection.cursor().execute(...) or via other sql-management editor
I think (may be I'm wrong), the best way is to save them in utf-8 encoding. But if someone wants, he can use another encoding.
I've added one more parameter "encoding" to the method as_string
### Solution 3
```
from psycopg2 import sql
import psycopg2.extensions


def as_string(composable, encoding="utf-8"):
    if isinstance(composable, sql.Composed):
        return ''.join([as_string(x) for x in composable])
    elif isinstance(composable, sql.SQL):
        return composable.string
    else:
        rv = sql.ext.adapt(composable._wrapped)
        if isinstance(rv, psycopg2.extensions.QuotedString): rv.encoding = encoding
        rv = rv.getquoted()
        return rv.decode(encoding) if isinstance(rv, bytes) else rv


# Testing...

query = sql.SQL("insert into mytable VALUES ({});") \
    .format(sql.SQL(', ').join(map(sql.Literal, ["any 'text' in 日本語 and even €", 1.2e-1, datetime.date.today()])))

print(as_string(query))
print(as_string(sql.Literal("1")))
print(as_string(sql.Literal(1e-2)))
print(as_string(sql.Literal(datetime.datetime.now())))
print(as_string(sql.Literal("ภาษาใดก็ได้")))
print(as_string(sql.Literal("Greek encoding: οποιαδήποτε γλώσσα"), "IBM869"))

s = as_string(sql.SQL("insert into mytable VALUES ({});").format(sql.Literal("€")))
if s != "insert into mytable VALUES ('€');": raise RuntimeError("Something went wrong")
# connection.cursor().execute(s)
# connection.cursor().execute("insert into mytable VALUES ('€');")
```
I don't see the difference (the last lines are examples). If it is impossible to insert "€" to the table because of database encoding, then it does not matter in what way it will be impossible,
via as_string(..)
`connection.cursor().execute(s)`
or directly
`connection.cursor().execute("insert into mytable VALUES ('€');")`

**Comment by nrogers-rew** [comment]
I should not need a database instance to determine whether SQL statements are _syntactically_ correct. 

I don't care when iterating (or unit testing) what my database encoding is—I do care whether what I am actually constructing semantically is what I expected.

Python solved the naive user problem in 2008 by making `bytes` objects incompatible with `str` objects. Just allow `context` to be an encoding? Be explicit in the documentation that you shouldn't do this in your code, just like you shouldn't construct arbitrary (plain) strings out of arbitrary data into SQL statements—a thing you are also completely free to do?

**Comment by jordanbreen28** [comment]
This is still a problem.

**Comment by dvarrazzo** [comment]
FYI this feature was implemented in psycopg 3.2. It will not be backported in psycopg 2.
